gnome-image-installer/util/gis-unattended-config.c

```diff
@@ -0,0 +1,175 @@
+ * 02111-1307, USA.
+ */
+#include "config.h"
+#include "gis-unattended-config.h"
+
+#include <glib/gi18n.h>
+
+#include "glnx-errors.h"
+
+typedef struct _GisUnattendedConfig {
+  GObject parent;
+
+  gchar *file_path;
+
+  GKeyFile *key_file;
+} GisUnattendedConfig;
+
+G_DEFINE_QUARK (gis-unattended-error, gis_unattended_error);
+
+static void gis_unattended_config_initable_iface_init (GInitableIface *iface);
+
+G_DEFINE_TYPE_EXTENDED (GisUnattendedConfig,
+                        gis_unattended_config,
+                        G_TYPE_OBJECT,
+                        0,
+                        G_IMPLEMENT_INTERFACE (G_TYPE_INITABLE,
+                                               gis_unattended_config_initable_iface_init));
+
+typedef enum {
+  PROP_FILE_PATH = 1,
+  N_PROPERTIES
+} GisUnattendedConfigPropertyId;
+
+static GParamSpec *props[N_PROPERTIES] = { 0 };
+
+static void
+gis_unattended_config_init (GisUnattendedConfig *self)
+{
+  self->key_file = g_key_file_new ();
+}
+
+static void
+gis_unattended_config_set_property (GObject      *object,
+                                    guint         property_id,
+                                    const GValue *value,
+                                    GParamSpec   *pspec)
+{
+  GisUnattendedConfig *self = GIS_UNATTENDED_CONFIG (object);
+
+  switch ((GisUnattendedConfigPropertyId) property_id)
+    {
+    case PROP_FILE_PATH:
+      g_free (self->file_path);
+      self->file_path = g_value_dup_string (value);
+      break;
+
+    case N_PROPERTIES:
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
+      break;
+    }
+}
+
+static void
+gis_unattended_config_dispose (GObject *object)
+{
+  GisUnattendedConfig *self = GIS_UNATTENDED_CONFIG (object);
+
+  g_clear_pointer (&self->key_file, g_key_file_unref);
+
+  G_OBJECT_CLASS (gis_unattended_config_parent_class)->dispose (object);
+}
+
+static void
+gis_unattended_config_finalize (GObject *object)
+{
+  GisUnattendedConfig *self = GIS_UNATTENDED_CONFIG (object);
+
+  g_clear_pointer (&self->file_path, g_free);
+
+  G_OBJECT_CLASS (gis_unattended_config_parent_class)->finalize (object);
+}
+
+static void
+gis_unattended_config_class_init (GisUnattendedConfigClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->set_property = gis_unattended_config_set_property;
+  // object_class->get_property = gis_unattended_config_get_property;
+  object_class->dispose = gis_unattended_config_dispose;
+  object_class->finalize = gis_unattended_config_finalize;
+
+  props[PROP_FILE_PATH] = g_param_spec_string (
+      "file-path",
+      "file path",
+      "Path to unattended.ini file",
+      NULL,
+      G_PARAM_WRITABLE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  g_object_class_install_properties (object_class, N_PROPERTIES, props);
+}
+
+static gboolean
+gis_unattended_config_initable_init (GInitable    *initable,
```

Comment:
Technically the implementation of `GInitable.init` needs to be idempotent, which means saving the `GError` or success value in your private struct.

Reply:
It's probably already (observably) idempotent as long as the underlying file doesn't change, though it does leak a few things. Ok.

Reply:
https://developer.gnome.org/gio/stable/GInitable.html#g-initable-init has the following:

> Callers should not assume that a class which implements GInitable can be initialized multiple times, unless the class explicitly documents itself as supporting this. Generally, a class’ implementation of init() can assume (and assert) that it will only be called once. Previously, this documentation recommended all GInitable implementations should be idempotent; that recommendation was relaxed in GLib 2.54.

so I will not change this, particularly since the code using this class does not assume it can be initialized multiple times.

Reply:
…maybe add a comment briefly explaining this? :-)

Reply:
I added an assertion, as suggested in the documentation above.